test/resolver_test.go

```diff
@@ -0,0 +1,194 @@
+	sendConfigChan := make(chan *iresolver.RPCConfig, 1)
+	gotContextChan := make(chan context.Context, 1)
+
+	ss := &stubServer{
+		emptyCall: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			gotContextChan <- ctx
+			return &testpb.Empty{}, nil
+		},
+	}
+	ss.r = manual.NewBuilderWithScheme("confSel")
+
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	state := iresolver.SetConfigSelector(resolver.State{
+		Addresses:     []resolver.Address{{Addr: ss.address}},
+		ServiceConfig: parseCfg(ss.r, "{}"),
+	}, funcConfigSelector{
+		f: func(i iresolver.RPCInfo) *iresolver.RPCConfig {
+			gotInfoChan <- &i
+			cfg := <-sendConfigChan
+			if cfg != nil && cfg.Context == nil {
+				cfg.Context = i.Context
+			}
+			return cfg
+		},
+	})
+	ss.r.UpdateState(state) // Blocks until config selector is applied
+
+	longdeadlineCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	shorterTimeout := 3 * time.Second
+
+	testMD := metadata.MD{"footest": []string{"bazbar"}}
+	mdOut := metadata.MD{"handler": []string{"value"}}
+
+	var onCommittedCalled bool
+
+	testCases := []struct {
+		name   string
+		md     metadata.MD
+		config *iresolver.RPCConfig
+	}{{
+		name:   "basic",
+		md:     testMD,
+		config: &iresolver.RPCConfig{},
+	}, {
+		name: "alter MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: metadata.NewOutgoingContext(ctx, mdOut),
+		},
+	}, {
+		name: "alter timeout; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: longdeadlineCtx,
+		},
+	}, {
+		name:   "nil config",
+		md:     metadata.MD{},
+		config: nil,
+	}, {
+		name: "alter timeout via method config; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			MethodConfig: serviceconfig.MethodConfig{
+				Timeout: &shorterTimeout,
+			},
+		},
+	}, {
+		name: "onCommitted callback",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			OnCommitted: func() {
+				onCommittedCalled = true
+			},
+		},
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			select {
+			case sendConfigChan <- tc.config:
+			default:
+				t.Fatalf("last config not consumed by config selector")
+			}
+
+			onCommittedCalled = false
+			ctx = metadata.NewOutgoingContext(ctx, tc.md)
+			if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+				t.Fatalf("client.EmptyCall(_, _) = _, %v; want _, nil", err)
+			}
+
+			var gotInfo *iresolver.RPCInfo
+			select {
+			case gotInfo = <-gotInfoChan:
+			default:
+				t.Fatalf("no config selector data")
+			}
+
+			var gotContext context.Context
+			select {
+			case gotContext = <-gotContextChan:
+			default:
+				t.Fatalf("no context received")
+			}
+
+			if want := "/grpc.testing.TestService/EmptyCall"; gotInfo.Method != want {
```

Comment:
Could we move these two checks to inside of the select case in line 145? I feel grouping this with the code where we read `gotInfo` would make it easier to follow whats happening in the test.